MIPX v1 - New json rpc method for swaps (#25)
MetaMask/metamask-improvement-proposals: MIPs/mip-3.md

---
MIP: 1
Title: Implement `wallet_swap` to support Dapps to swap
Status: Draft
Stability: Experimental
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
Author(s): Tomás Santos (@tommasini)
Type: Community
Created: 2023-11-10
---

## Summary

This proposal introduces a new JSON RPC method `wallet_swap`` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.

## Motivation

The motivation behind this MIP is to provide a standardized way for dApps to request token swaps from MetaMask, improving interoperability and user experience.

# Usage Example

```

> ethereum.request({
>   method: 'wallet_swap',
>   params: {
>     from: {
>       token_address: '0x1234567890abcdefABCDEF1234567890ABCDEF',
>       chainId: '0x1',
>       amount: '0xDE0B6B3A7640000',
>     },
>     to: {
>       token_address: '0xabcdef1234567890ABCDEF1234567890abcdef',
>       chainId: '0x1',
>     },
>   },
> });

Reviewer: we should be using [`by-position` params](https://json-rpc.org/specification#parameter_structures). The only method that is `by-name` is wallet_watchAsset and that was a mistake that it went out like that.

```suggestion
ethereum.request({
  method: 'wallet_swap',
  params: [
    {
      from: {
        token_address: '0x1234567890abcdefABCDEF1234567890ABCDEF',
        chainId: '0x1',
        amount: '0xDE0B6B3A7640000',
      },
      to: {
        token_address: '0xabcdef1234567890ABCDEF1234567890abcdef',
        chainId: '0x1',
      },
    },
  ]
});
```

Author: Interesting! Will do this changes! Thank you so much!

- Increase revenue - Easier adoption means more users will use MetaMask swaps.

Introducing this API directly in the wallet is preferable when compared with the alternative of using snaps, as the latter would create more friction and less adoption since users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.

Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.

# Usage Example

```
ethereum.request({
  method: 'wallet_swapAsset',
  params: [{
    from: [{
      tokenAddress: '0x1234567890abcdefABCDEF1234567890ABCDEF',
      value: '0xDE0B6B3A7640000',
    }],
    to: {
      tokenAddress: '0xabcdef1234567890ABCDEF1234567890abcdef',
    },
    userAddress: '0x0000000000000000000000000000000000000000'
  },
  ]
});

```

## Proposal Specification

The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:

- `from`: An object containing details about the source token. It should include:

  - `tokenAddress`: The address of the source token.

>   - `value`: The amount of wei in hexadecimal format of the source token to be swapped.

Author: Change the value to be optional! Default to 0 if it's not defined.

Author: Added on this [commit ](https://github.com/MetaMask/metamask-mobile/pull/7509/commits/6f675ad1110107894271f2e16f5b5ae98e6ee187) still need to introduce it on the MIP

---

> MIP: 3

Author: Should change the name of the file also to mip-x?

> MIP: 3

Reviewer 2: Yes also should be mip-x.md

> MIP: 3

Author: https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/2b1211af223fe5f926868f958b67d6a5065a2533

Addressed on this commit

Title: Implement `wallet_swapAsset` to support Dapps to swap
Status: Draft
Stability: Experimental
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/46
Author(s): Tomás Santos (@tommasini)
Type: Community
Created: 2023-11-10
---

## Summary

> This proposal introduces a new JSON RPC method `wallet_swapAsset` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.

Reviewer 3: nit: `JSON RPC` should be connected by a dash. `JSON-RPC`

Author: Thanks!!! Replaced everywhere on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/55c71d31a08464e00b9a3cfb055d09f17eca025d)

## Motivation

> This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.

Reviewer 3: ```suggestion
This proposal aims to solve a key pain point for dApp developers and users - the inability to initiate swap tokens from the dApp interface.
```

the swap is fully contained within the wallet after initiated by the dApp

Author: Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.

> Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:

Reviewer 3: same nit here about wording. The swap doesn't happen inside the dApp, it's only initiated from the dApp

Author: Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

> - Improved usability - Users can get the right tokens for dApp operations in fewer steps.

Reviewer 3: ```suggestion
- Improved usability - Users can get the right tokens for dApp operations in fewer steps and with less context switching.
```

maybe? idk. feel free to ignore this suggestion

Author: Totally!
Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
- Interoperability - Standard method works across different wallets.
- Increase revenue - Easier adoption means more users will use MetaMask swaps.

> Comparing with other alternative, implementing it using swaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The all goal is to reduce friction to swaps.

Reviewer 4: `implementing it using swaps` - Was this intended to say `snaps`?
`The all goal is to reduce friction to swaps.` - Can simplify to `The goal...`

> Introducing this API directly in the wallet is preferable when compared with the alternative of using snaps, as the latter would create more friction and less adoption since users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.

Reviewer 3: to play devil's advocate here, is there a world in which a snap that is able to facilitate swapping is installed and enabled by default for all metamask users?

Author: Yeah, I wrote this some time ago! I agree with you, snaps have a lot to grow, and probably ti can appear something that could bring better UX (which would be awesome!) I'm unfamiliar with the snaps possibilities (Really need more time to explore it and want to do it in the second semester of this year). Do you think this json-rpc method used on the snap facilitates that to happen?

Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.

# Usage Example

```
ethereum.request({
  method: 'wallet_swapAsset',
  params: [{
    from: [{
      tokenAddress: '0x1234567890abcdefABCDEF1234567890ABCDEF',
      value: '0xDE0B6B3A7640000',
    }],
    to: {
      tokenAddress: '0xabcdef1234567890ABCDEF1234567890abcdef',
    },
    userAddress: '0x0000000000000000000000000000000000000000'
  },
  ]

> });

Reviewer 2: Just a thought should we put the `userAddress` in the `from` object? And then optionally the sendToAddress in the `to` object?

> ethereum.request({
>   method: 'wallet_swapAsset',
>   params: [{
>     from: [{
>       tokenAddress: '0x1234567890abcdefABCDEF1234567890ABCDEF',
>       value: '0xDE0B6B3A7640000',
>     }],
>     to: {
>       tokenAddress: '0xabcdef1234567890ABCDEF1234567890abcdef',
>     },
>     userAddress: '0x0000000000000000000000000000000000000000'
>   },
>   ]
> });

Reviewer 2: Yeah I can see that. Though I'm not sure that `from` and `to` are so obviously only toke related? We could name them `fromToken` and `toToken`:
```suggestion
ethereum.request({
  method: 'wallet_swapAsset',
  params: [{
    fromToken: [{
      address: '0x1234567890abcdefABCDEF1234567890ABCDEF',
      value: '0xDE0B6B3A7640000',
    }],
    toToken: {
      address: '0xabcdef1234567890ABCDEF1234567890abcdef',
    },
    userAddress: '0x0000000000000000000000000000000000000000'
  },
  ]
});
```

> });

Author: Like it! Much more self explanatory <3

> });

Author: Addressed on this commit: 
https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/2b1211af223fe5f926868f958b67d6a5065a2533

```

## Proposal Specification

The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:

- `from`: An object containing details about the source token. It should include:

>   - `token_address`: The address of the source token.

Reviewer 4: Should we specify format here? For example, checksummed address?

Reviewer 5: The openRPC spec may be the best place to specify formats.
Here's doc example with some formatting guidelines included for some params: https://docs.metamask.io/wallet/reference/personal_sign/
And in the api-spec: https://github.com/MetaMask/api-specs/blob/6d55565c9312923b533e580efca88f98364c4da4/openrpc.yaml#L1074

>   - `amount`: The amount on wei in hexadecimal format of the source token to be swapped.

Reviewer: ```suggestion
  - `value`: The amount of wei in hexadecimal format of the source token to be swapped.
```

should use `value` here to match `eth_sendTransaction` as its wei there too to keep the api consistent

Author: Addressed! Thanks!!

>   - `amount`: The amount on wei and hexadecimal format of source token to be swapped.

Reviewer 4: ```suggestion
  - `amount`: The amount of wei in hexadecimal format of the source token to be swapped.
```

- `to`: An object containing details about the destination token. It should include:

  - `token_address`: The address of the destination token.

> - `user_address`: An string containing the address connected to the dapp.

Reviewer: can we use camel case for keys like we are doing here with `sendTransaction` with `gasPrice` etc:

![image](https://github.com/MetaMask/metamask-improvement-proposals/assets/364566/6823eda8-7812-480f-bcbe-094196157b6e)

Author: Definitely!

> - `sendTo`: (Optional) Allows integration with Swap's `send to` feature.

Reviewer 3: i don't like this field as it overloads swap into also being send. Even if MetaMask swap supports it, I don't believe we should expose it as part of this method

> - `sendTo`: (Optional) Allows integration with Swap's `send to` feature.

Author: @jiexi thanks for the comment! Could you elaborate on "overloads swap into also being send"? I'm not sure what you mean with it!

> - `sendTo`: (Optional) Allows integration with Swap's `send to` feature.

Reviewer 2: > Even if MetaMask swap supports it, I don't believe we should expose it as part of this method

Could you elaborate on why you think the the API shouldn't support it even if the wallet has moved to support it?

> - `sendTo`: (Optional) Allows integration with Swap's `send to` feature.

Author: After conversation, we decided to have the sendToAddress as a reference for future implementation on the MIP and remove it from the api spec

> - `referall_code`: (Optional) Future implementation will allow adding a referall code to bigger adoption of the json rpc method across the builders community

Reviewer 4: ```suggestion
- `referral_code`: (Optional) Future implementation will allow adding a referral code to drive adoption of the JSON RPC method across the builders community
```
Is it meant to be referral code?

Author: Yes1!! Thanks!!! Great catch !!

> - `send_to`: (Optional) Allows integration of the implementation of the swap send to feature.

Reviewer 4: ```suggestion
- `send_to`: (Optional) Allows integration with Swap's `send to` feature.
```
Does the suggestion mean the same?

Author: Yess!!!

MetaMask will interpret the method call and perform the necessary validations and operations to initiate the token swap.

## Caveats

> The implementation of `wallet_swapAsset` will be new and this could have some time for adoption.

Reviewer 4: ```suggestion
The implementation of `wallet_swapAsset` will be new and may take some time to be adopted.
```

Reviewer 2: I'm not sure this caveat is necessary. Ramp up time for adoption can be assumed.

Author: Yeah, I aggree, removed!

1. Backward compatibility: The introduction of the `wallet_swapAsset` method does not affect existing methods or functionality. Developers can integrate this method into their dApps without needing to modify existing code.

2. Ease of adoption: The `wallet_swapAsset` method is designed to be straightforward to use. Developers simply need to provide the necessary parameters (source token details, destination token details) in the method call. The wallet handles the rest of the process, including validation, token swap operation, and error handling.

3. User Experience: By using the `wallet_swapAsset` method, developers can improve the user experience of their dApps. Users can perform token swaps without leaving the dApp, making the process more seamless and convenient.

4. Error handling: Dapp Developers should implement appropriate error handling when using the `wallet_swapAsset` method. The method can return various errors, such as invalid parameters or unsupported operations, and these should be handled gracefully in the dApp.
   On MetaMask side we intent to validate every property of the request and return an approriate answer to the dapp.

>    Currently we throw these errors:

Reviewer 3: this needs to be filled out still, perhaps in a separate section dedicated just to result and errors

4. Error handling: Dapp Developers should implement appropriate error handling when using the `wallet_swapAsset` method. The method can return various errors, such as invalid parameters or unsupported operations, and these should be handled gracefully in the dApp.
   On MetaMask side we intent to validate every property of the request and return an approriate answer to the dapp.

>    Currently we throw this errors:

Reviewer 4: ```suggestion
   Currently we throw these errors:
```

> - **Undefined Parameters**: Parameters like `from`, `to`, `user_address`, etc. are essential for the `wallet_swapAsset` method to work correctly. If not provided, the `validateParams` function would throw an error saying `"${property} property of ${name} is not defined"`. (This validation will change on the future when we allow multiple swap tokens, since the architecture of this rpc method allows it.)

Reviewer 4: ```suggestion
- **Undefined Parameters**: Parameters like `from`, `to`, `user_address`, etc. are essential for the `wallet_swapAsset` method to work correctly. If not provided, the `validateParams` function would throw an error saying `"${property} property of ${name} is not defined"`. (This validation will change in the future when we allow multiple swap tokens, since the architecture of this RPC method allows it.)
```

> 5. Ensuring Correct Chain Context for wallet_swapAsset: Before initiating a swap, it's crucial to ensure that the user's wallet is connected to the expected blockchain network. Developers should use the [eth_chainId](https://docs.metamask.io/wallet/reference/eth_chainid/) method to verify the current chain, [wallet_switchEthereumChain](https://docs.metamask.io/wallet/reference/wallet_switchethereumchain/) to switch to the desired chain if necessary, and [wallet_addEthereumChain](https://docs.metamask.io/wallet/reference/wallet_addethereumchain/) to add a new chain if it's not known to the wallet. This ensures that token swaps are performed on the intended network, enhancing security and user experience.

Reviewer 3: ```suggestion
5. Ensuring Correct Chain Context for `wallet_swapAsset`: Before initiating a swap, it's crucial to ensure that the user's wallet is connected to the expected blockchain network. Developers should use the [eth_chainId](https://docs.metamask.io/wallet/reference/eth_chainid/) method to verify the current chain, [wallet_switchEthereumChain](https://docs.metamask.io/wallet/reference/wallet_switchethereumchain/) to switch to the desired chain if necessary, and [wallet_addEthereumChain](https://docs.metamask.io/wallet/reference/wallet_addethereumchain/) to add a new chain if it's not known to the wallet. This ensures that token swaps are performed on the intended network, enhancing security and user experience.
```

Author: Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

> - **Undefined Parameters**: Parameters like `from`, `to`, `userAddress`, etc. are essential for the `wallet_swapAsset` method to work correctly. If not provided, the `validateParams` function would throw an error saying `"${property} property of ${name} is not defined"`. (This validation will change on the future when we allow multiple swap tokens, since the architecture of this rpc method allows it). See more details about error codes in the api-spec: https://github.com/MetaMask/api-specs/pull/201

Reviewer 3: `Required Parameters`?

Author: Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

  ```markdown
  validateParams(from[0], [amount, token_address], 'from');
  validateParams(to, [token_address], 'to');
  ```

- **Invalid User Address**: If a non-existent address is provided in `user_address`, then the method will throw an error 'This address does not exist'.

  ```markdown
  if (!dappConnectedAccount) {
  throw ethErrors.rpc.invalidParams('This address does not exist');
  }
  ```

> - **Unsupported Multiple Tokens Swap**: As of now, it supports only a single token swap. In case of multiple tokens provided in the `from` object, an 'Currently we de not support multiple tokens swap' error is thrown.

Reviewer 4: ```suggestion
- **Unsupported Multiple Tokens Swap**: As of now, it supports only a single token swap. In case of multiple tokens provided in the `from` object, an 'Currently we do not support multiple tokens swap' error is thrown.
```

>   validateParams(from[0], [tokenAddress], 'from');

Reviewer 2: Oh here it looks like from is also expected to be an array?

  ```markdown
  if (from.length > 1) {
  throw ethErrors.rpc.invalidParams(

>   'Currently we de not support multiple tokens swap',

Reviewer 4: ```suggestion
  'Currently we do not support multiple tokens swap',
```

  );
  }
  ```

> - **Inactive or Unsupported Swaps**: If the swap is inactive or not possible on the current chain, an alert with the message 'Swap is not active or not possible on this chain' will be triggered.

Reviewer 4: ```suggestion
- **Inactive or Unsupported Swaps**: If the swap is inactive or is not possible on the current chain, an alert with the message 'Swap is not active or is not possible on this chain' will be triggered.
```

  ```markdown
  const isSwappable = isSwapsAllowed(chainId) && swapsIsLive;
  if (!isSwappable) {
  Alert.alert('Swap is not active or not possible on this chain');
  return;
  }
  ```

> 5. Security: Dapp Developers should ensure that the token addresses and chain IDs provided in the method call are valid and secure. They should also inform users about the potential risks and considerations when performing token swaps.

Reviewer 5: ```suggestion
5. Security: Dapp Developers should ensure that the token addresses provided in the method call are secure and valid for the current chain ID. They should also inform users about the potential risks and considerations when performing token swaps.
```

>   }

Author: Yeah, It was a while back but I talked I remember talking with Forest about it!

Read top to bottom so answered to you on top as well ahah

>   }

Author: Raised the question on slack for us to be sure: https://consensys.slack.com/archives/C06S05257NX/p1717078087741209

>    On MetaMask platfors we have warnings in place that say to the user to be carefull when a token is not that trustworthy. We check if the number of occurrences is more than one, If not we show an alert the we obligate the user to ready and press a button to be able to swap.

Reviewer 4: ```suggestion
   On MetaMask platforms, we have warnings in place that communicates to the user to be careful when a token is not trustworthy. We check if the number of occurrences is more than one. If it isn't, we then prompt and alert to the user, which requires a user to acknowledge before they can continue with the swap.
```

Does that last part keep the same meaning?

> 6. Testing: Before deploying the `wallet_swapAsset` method in a live environment, MetaMask engineers should thoroughly test it to ensure it works correctly and handles errors appropriately. This includes testing with different token types, amounts, and network conditions.

Reviewer 4: ```suggestion
6. Testing: Before deploying the `wallet_swapAsset` method in a live environment, MetaMask engineers should thoroughly test it to ensure the it works correctly and handles errors appropriately. This includes testing with different token types, amounts, and network conditions.
```

Author: Not sure if this makes sense

7. Interoperability: The `wallet_swapAsset` method provides a standardized way for dApps to request token swaps from wallets. This can improve interoperability between different dApps and wallets, as they can all use the same method for token swaps.

## User Experience Considerations

Users will now be able to swap tokens without leaving dapps.

> The `wallet_swapAsset` method might be new to some users, providing educational resources can help them understand how it works and how to use it.

Reviewer 4: ```suggestion
Since `wallet_swapAsset` method will be new to some users, providing educational resources may help in understand how the method works as well as how to use it.
```

> 5. Security: Dapp Developers should ensure that the token addresses provided in the method call are secure and valid for the current chain ID. They should also inform users about the potential risks and considerations when performing token swaps.
>    On MetaMask platforms we have warnings in place that say to the user to be carefull when a token is not that trustworthy. We check if the number of occurrences is more than one, If not we show an alert the we obligate the user to ready and press a button to be able to swap.

Reviewer 3: ```suggestion
5. Security: In MetaMas we have warnings in place to alert the user when a token is not trustworthy and add an additional confirmation step. Despite this, dApp developers should still ensure that the token addresses provided in the method call are secure and valid for the current chain ID. They should also inform users about the potential risks and considerations when performing token swaps.
```

maybe?

Author: Sounds good!

Author: Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)

If the `wallet_swapAsset` method encounters an error, the wallet/dapp should provide a clear and understandable error message to the user. This helps users understand what went wrong and how to fix it.

> On MetaMask Mobile, if the user have on the wallet an selected account but on the dapp in app browser another, when using swaps feature, MetaMask Mobile will check if the accounts are different and switch to the account selected on the dapp if they are.

Reviewer 4: ```suggestion
While using MetaMask Mobile with this method, the mobile app will first check if selected account is different between the wallet and the dApp. If it is different, the mobile app will switch to the matching account before continuing with the swap.
```

Author: Changed! Thanks!

```markdown
if (
safeToChecksumAddress(selectedAddress) !==
checksummedDappConnectedAccount
) {
Engine.context.PreferencesController.setSelectedAddress(
checksummedDappConnectedAccount,
);
}
```

## Security Considerations

> While the proposed implementation of `wallet_swapAsset` do not introduce new security risks in terms of contract interactions, they do present a potential risk related to phishing, althought it is good to mention that this risk already exist on our swaps flow.

Reviewer 4: ```suggestion
While the proposed implementation of `wallet_swapAsset` does not introduce new security risks in terms of contract interactions, they do present a potential risk related to phishing.
```

Not sure if we should mention that last part `althought it is good to mention that this risk already exist on our swaps flow.`

Author: yeah will remove it!! I don't think it should me mentioned as well!

> Malicious actors may prompt users to swap fake tokens that they created and fake on the dapp that the users are doing swap to an official token.

Reviewer 4: ```suggestion
Malicious actors may prompt users to swap fake tokens and convince users that they are performing a swap with an official token.
```

Author: Changed!!!! Thanks <3

> To mitigate this risk, MetaMask have implemented the following countermeasure:

Reviewer 4: ```suggestion
To mitigate this risk, MetaMask has implemented the following countermeasure:
```

> ```

Reviewer 2: Is there a reason we are saying we'll only do this in mobile? I guess is it not required in extension?

Author: On mobile it's needed because you can have a selected account on the wallet view and a different selected account on the dapp on in-app browser, and what json rpc does now is redirect the user to the swaps flow, so if the address between the wallet view and in app browser dapp is different, we change the account on the wallet view to the one at the dapp.

On extension this doesn't happen.

Author: Added here: https://github.com/tommasini/metamask-improvement-proposals/blob/e3f72449345cddb460f76c60df53a140b1560c0d/MIPs/mip-3.md#L157

**User Education:** Inform users about the potential risks associated with token swaps and provide guidelines for safely performing these operations. Users should be advised to verify the authenticity of the tokens they are swapping and the dApps they are using.

Developers and users should also be aware of the standard security considerations when dealing with tokens, such as verifying contract addresses and ensuring the trustworthiness of token issuers.

## References

[EIP: wallet_swapAsset JSON RPC method](https://github.com/tommasini/EIPs/blob/64090fe54761906fc9bde31ad46eb85230de5e8d/EIPS/eip-swap-rpc.md)

## Feedback

Please provide feedback on this proposal by opening an issue in the MetaMask MIPs repository.

## Committed Developers

> Tomás Santos (@tommasini)

Reviewer 5: Is this something the Swaps team might be interested in adopting to streamline interactions from the Portfolio dapp's [swap page](https://portfolio.metamask.io/swap) to the wallet?

Author: That's an amazing question! Let me follow up on this with you via DM

> Tomás Santos (@tommasini)

Reviewer 4: 👏

## Motivation

This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.

Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.

Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:

- Improved usability - Users can get the right tokens for dApp operations in fewer steps.
- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
- Interoperability - Standard method works across different wallets.
- Increase revenue - Easier adoption means more users will use MetaMask swaps.

> Comparing with other alternative, implementing it using snaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.

Reviewer 2: ```suggestion
Introducing this API directly in the wallet is preferable when compared with the alternative of using snaps, as the latter would create more friction and less adoption since users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.
```

Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.

# Usage Example

```
ethereum.request({
  method: 'wallet_swapAsset',
  params: [{
    from: [{
      tokenAddress: '0x1234567890abcdefABCDEF1234567890ABCDEF',
      value: '0xDE0B6B3A7640000',
    }],
    to: {
      tokenAddress: '0xabcdef1234567890ABCDEF1234567890abcdef',
    },
    userAddress: '0x0000000000000000000000000000000000000000'
  },
  ]
});

```

## Proposal Specification

The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:

- `from`: An object containing details about the source token. It should include:

  - `tokenAddress`: The address of the source token.
  - `value`: The amount of wei in hexadecimal format of the source token to be swapped.

- `to`: An object containing details about the destination token. It should include:

  - `tokenAddress`: The address of the destination token.

- `userAddress`: Account address connected to the dapp.

> - `referralCode`: (Optional) Future implementation will allow adding a refferal code to drive adoption of the JSON RPC method across the builders community.

Reviewer 2: ```suggestion
- `referralCode`: (Optional) Future enhancements to this API will allow adding a referral code by which referrers can earn a percent of swap fees associated with the swaps they propose. This will help to further drive adoption of this API.
```

- `sendTo`: (Optional) Allows integration with Swap's `send to` feature.

MetaMask will interpret the method call and perform the necessary validations and operations to initiate the token swap.

## Caveats

The implementation of `wallet_swapAsset` will be new and this could have some time for adoption.

> ## Implementation

Reviewer 2: this section could probably use a bit more explanation of what specific work will be required to enable this feature.
